pkg/file/redundancy/getter/getter.go

```diff
@@ -0,0 +1,226 @@
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/klauspost/reedsolomon"
+)
+
+// decoder is a private implementation of storage.Getter
+// if retrieves children of an intermediate chunk potentially using erasure decoding
+// it caches sibling chunks if erasure decoding started already
+type decoder struct {
+	fetcher    storage.Getter  // network retrieval interface to fetch chunks
+	putter     storage.Putter  // interface to local storage to save reconstructed chunks
+	addrs      []swarm.Address // all addresses of the intermediate chunk
+	inflight   []atomic.Bool   // locks to protect wait channels and RS buffer
+	cache      map[string]int  // map from chunk address shard position index
+	waits      []chan struct{} // wait channels for each chunk
+	rsbuf      [][]byte        // RS buffer of data + parity shards for erasure decoding
+	ready      chan struct{}   // signal channel for successful retrieval of shardCnt chunks
+	shardCnt   int             // number of data shards
+	parityCnt  int             // number of parity shards
+	wg         sync.WaitGroup  // wait group to wait for all goroutines to finish
+	mu         sync.Mutex      // mutex to protect buffer
+	fetchedCnt atomic.Int32    // count successful retrievals
+	cancel     func()          // cancel function for RS decoding
+	remove     func()          // callback to remove decoder from decoders cache
+}
+
+type Getter interface {
+	storage.Getter
+	io.Closer
+}
+
+// New returns a decoder object used tos retrieve children of an intermediate chunk
+func New(addrs []swarm.Address, shardCnt int, g storage.Getter, p storage.Putter, strategy Strategy, strict bool, fetchTimeout time.Duration, remove func()) Getter {
+	ctx, cancel := context.WithCancel(context.Background())
+	size := len(addrs)
+	if fetchTimeout == 0 {
+		fetchTimeout = 30 * time.Second
+	}
+	strategyTimeout := StrategyTimeout
+
+	rsg := &decoder{
+		fetcher:   g,
+		putter:    p,
+		addrs:     addrs,
+		inflight:  make([]atomic.Bool, size),
+		cache:     make(map[string]int, size),
+		waits:     make([]chan struct{}, shardCnt),
+		rsbuf:     make([][]byte, size),
+		ready:     make(chan struct{}, 1),
+		cancel:    cancel,
+		remove:    remove,
+		shardCnt:  shardCnt,
+		parityCnt: size - shardCnt,
+	}
+
+	// after init, cache and wait channels are immutable, need no locking
+	for i := 0; i < shardCnt; i++ {
+		rsg.cache[addrs[i].ByteString()] = i
+		rsg.waits[i] = make(chan struct{})
+	}
+
+	// prefetch chunks according to strategy
+	rsg.wg.Add(1)
+	go func() {
+		rsg.prefetch(ctx, strategy, strict, strategyTimeout, fetchTimeout)
+		rsg.wg.Done()
+	}()
+	return rsg
+}
+
+// Get will call parities and other sibling chunks if the chunk address cannot be retrieved
+// assumes it is called for data shards only
+func (g *decoder) Get(ctx context.Context, addr swarm.Address) (swarm.Chunk, error) {
+	i, ok := g.cache[addr.ByteString()]
+	if !ok {
+		return nil, storage.ErrNotFound
+	}
+	if g.fly(i, true) {
```

Comment:
the `prefetch` calls essentially sets all of these chunks to inflight, no?
why is this fly check necessary, why can't we jump to the select below?

Reply:
because we want singleflight behaviour on fetching, ie if prefetching fetches a chunk,, then its queried with joiner Get --> decoder Get (or the other way round), then we should just wait on the inflight fetch.
Similarly, when prefetch fetched shardCnt chunks the other chunks can be put to inflight, so that if they are `Get-ed' by the joiner they are not fetched but wait till revovered.

Reply:
And of course there is not always a prefetch on every data chunk (NONE and PROX will not prefetch some chunks)